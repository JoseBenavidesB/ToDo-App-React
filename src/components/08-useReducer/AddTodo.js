import React from 'react'
import { useForm } from '../../hooks/useForm';

export const AddTodo = ({ handleAddToDo }) => {

    const [ { description }, handleInputChange, reset ] = useForm({
        description: '',

    });

    const handleSubmit = (e) => {
        e.preventDefault();
        if ( description.trim().length <= 1) { return;}

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };
        
        handleAddToDo(newTodo);
        reset();
    }
    return (
        <>
            <h4> Add ToDo </h4> <hr/>

            <form onSubmit={ handleSubmit }>
                
                <input
                    type="text"
                    name="description"
                    className='form-control mb-0 m-auto'
                    placeholder="Write something"
                    autoComplete="off"
                    value={ description }
                    onChange={ handleInputChange }
                />

                <button 
                    type= "submit" 
                    className='btn btn-outline-primary mt-2 col-12 m-auto'> Add </button>

            </form>
        </>
    )
}
