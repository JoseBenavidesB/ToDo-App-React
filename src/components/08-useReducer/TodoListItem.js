import React from 'react'

export const TodoListItem = ({ todo, idx , handleToggle, handleDelete }) => {
  return (
    <li key={todo.id} className="list-group-item">
      <p
        onClick={() => handleToggle(todo.id)}
        className={`${todo.done && "complete"}`}
      >
        {idx + 1}. {todo.desc}
      </p>

      <button
        // id= { todo.id }
        onClick={() => handleDelete(todo.id)}
        className="btn btn-danger m-2"
      >
        Delete
      </button>
    </li>
  );
}
