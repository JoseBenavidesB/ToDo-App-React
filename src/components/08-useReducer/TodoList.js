
import React from 'react'
import { TodoListItem } from './TodoListItem';

export const TodoList = ({ toDos = [], handleToggle, handleDelete }) => {
  return (
    <ul className="list-group list-group-flush p-0">
        {toDos.map((todo, idx) => (
            <TodoListItem key={todo.id} todo={ todo } idx={ idx }  handleDelete={handleDelete} handleToggle={handleToggle}/>
        ))}
    </ul>
    
  );
}
