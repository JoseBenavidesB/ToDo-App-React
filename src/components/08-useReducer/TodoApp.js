
import React from 'react'; 
import { useReducer, useEffect } from 'react';
import { todoReducer } from './todoReducer';
//import { useForm } from '../../hooks/useForm';


import './styles.css'
import { TodoList } from './TodoList';
import { AddTodo } from './AddTodo';


const init = ()=> {

    return JSON.parse(localStorage.getItem( 'todos')) || [];
    /* return [{
        id: new Date().getTime(),
        desc: 'Learn React',
        done: false
    }]; */
}
export const TodoApp = () => {

    const [toDos, dispatch ] = useReducer(todoReducer, [], init);

    useEffect(() => { //save in the local storage
        
        localStorage.setItem( 'todos', JSON.stringify( toDos ) )
      }
    , [toDos]);
    
    const handleDelete = ( id ) => {
        const action = {
            type: 'delete',
            payload: id
        }
        //dispatch
        dispatch(action);
    };

    const handleToggle = ( todoId ) => {
        dispatch( {
            type: 'toggle',
            payload: todoId
        })
    };
    const handleAddToDo = (newTodo) => {
        dispatch({
            type : 'add',
            payload : newTodo
        });
    };

    return (
        <>
            <h1> ToDo App ({ toDos.length })</h1> <hr/>

            <div key={1} className="row">
                <div className="col-7">

                    <TodoList 
                        toDos={ toDos } 
                        handleDelete={handleDelete} 
                        handleToggle={handleToggle} 
                    />

                </div >
                <div  className="col-5">
                    < AddTodo handleAddToDo={handleAddToDo}/>
                </div>
            </div>
            
        </>
    )
}
