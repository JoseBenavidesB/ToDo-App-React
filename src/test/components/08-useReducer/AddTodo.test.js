import { shallow } from 'enzyme/build'
import React from 'react'
import { AddTodo } from '../../../components/08-useReducer/AddTodo'


describe('Test AddTodo', () => {
    const handleAddToDo = jest.fn();

    const wrapper = shallow( <AddTodo handleAddToDo={ handleAddToDo }/>)

    test('should show correctly', () => { 
        
        expect( wrapper ).toMatchSnapshot();
     });

     test('dont call handleAddToDo', () => { 
        
        const formSubmit = wrapper.find('form').prop('onSubmit');

        formSubmit({ preventDefault(){} });

        expect( handleAddToDo ).toHaveBeenCalledTimes(0);

     });

     test('should call handleAddToDo', () => { 

        const value = 'learn react'

        wrapper.find('input').simulate('change', { 
            target: {
                value,
                name: 'description'
            }
        });
        
        const formSubmit = wrapper.find('form').prop('onSubmit');

        formSubmit({ preventDefault(){} });

        expect( handleAddToDo ).toHaveBeenCalledTimes(1);
        expect( handleAddToDo ).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: value,
            done: false,
        });
        
        expect( wrapper.find('input').prop('value') ).toBe('');//this is the RESET()

     });


});