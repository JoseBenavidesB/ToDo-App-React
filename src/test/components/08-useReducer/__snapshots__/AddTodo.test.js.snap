// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test AddTodo should show correctly 1`] = `
<Fragment>
  <h4>
     Add ToDo 
  </h4>
   
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control mb-0 m-auto"
      name="description"
      onChange={[Function]}
      placeholder="Write something"
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-2 col-12 m-auto"
      type="submit"
    >
       Add 
    </button>
  </form>
</Fragment>
`;
